files/tk_metrics

```diff
@@ -7,44 +7,69 @@ require 'time'
 require 'optparse'
 require 'yaml'
 
-options = {}
-OptionParser.new do |opts|
-  opts.banner = "Usage: tk_metrics [options]"
-
-  opts.on('-p', '--[no-]print', 'Print to stdout') { |p| options[:print] = p }
-  opts.on('-m [TYPE]', '--metrics_type [TYPE]', 'Type of metric to collect') { |v| options[:metrics_type] = v }
-  opts.on('-o [DIR]', '--output-dir [DIR]', 'Directory to save output to') { |o| options[:output_dir] = o }
-  opts.on('--metrics_port [PORT]', 'The port the metrics service runs on') { |port| options[:metrics_port] = port }
-  opts.on('--[no-]ssl', 'Whether or not to use SSL when gather metrics') { |ssl| options[:ssl] = ssl }
-end.parse!
+#===========================================================================#
+# BUILD CONFIGURATION OPTIONS                                               #
+#===========================================================================#
 
-if options[:metrics_type].nil? then
-  STDERR.puts '--metrics_type (-m) is a required argument'
-  exit 1
-end
+settings = { }
+config = { }
+
+OptionParser.new do |parser|
+  parser.banner = "Usage: tk_metrics [options]"
+
+  save_ident = lambda { |name,arg| config[name] = arg }
+  save_list  = lambda { |name,arg| config[name] = arg.split(',') }
 
-METRICS_TYPE = options[:metrics_type]
-config = YAML.load_file(File.join(File.dirname(File.expand_path(__FILE__)),"#{METRICS_TYPE}_config.yaml"))
+  opt_bool  = lambda { |arg| "--[no-]#{arg}" }
+  opt_value = lambda { |arg| "--#{arg} ARG" }
 
-def coalesce ( opt1, opt2, default=nil )
-  if opt1.nil? then
-    if opt2.nil? then
-      default
-    else
-      opt2
+  option = lambda do |name,desc,fopt,fsave,default=nil|
+    settings[name] = { default: default }
+    parser.on(fopt.call(name.to_s), desc) do |arg|
+      fsave.call(name, arg)
```

Comment:
I think it's just a difference between a functional style and a procedural style. I like the functional style, but I'm sure a procedural equivalent could be done too. I originally chose functional because I wanted to clearly contain all of the logic for how to build out config options (starts on line 33) INSIDE the `OptionParser#new` block, not rely on any external `def`'s, and not do any weird scoped `def`'s inside the block. Since I was doing one lambda anyway I figured I'd go all-in on the functional.